packages/lexicon/src/identifiers.ts

```diff
@@ -0,0 +1,317 @@
+// These are relatively permissive validators for identifier strings. The idea
+// is to have simple implementations here, and test coverage, that would be
+// enforced while doing Lexicon-level validation. These are redundant with
+// existing packages like 'handle', 'nsid', and 'uri'... but those packages
+// might be enforcing additional validation or constratins beyond those here.
+//
+// For use in consistent cross-language / cross-implementation validation, I
+// think it is beneficial to have these be simple and not rely on external
+// libraries.
+
+// Handle constraints, in English:
+//  - must be a possible domain name
+//    - RFC-1035 is commonly referenced, but has been updated. eg, RFC-3696,
+//      section 2. and RFC-3986, section 3. can now have leading numbers (eg,
+//      4chan.org)
+//    - "labels" (sub-names) are made of ASCII letters, digits, hyphens
+//    - can not start or end with a hyphen
+//    - TLD (last component) should not start with a digit
+//    - can't end with a hyphen (can end with digit)
+//    - must be between 2 and 63 characters (not including any periods)
+//    - overall length can't be more than 253 characters
+//    - separated by (ASCII) periods; does not start or end with period
+//    - case insensitive
+//    - domains (handles) are equal if they are the same lower-case
+//    - punycode allowed for internationalization
+//  - no whitespace, null bytes, joining chars, etc
+//  - does not validate whether domain or TLD exists, or is a reserved or
+//    special TLD (eg, .onion or .local)
+//  - does not validate punycode
+
+// This longer handle verifier gives understandable errors
+export const lexVerifyHandle = (handle: string): void => {
+  // check that all chars are boring ASCII
+  if (!/^[a-zA-Z0-9.-]*$/.test(handle)) {
+    throw new Error(
+      'Disallowed characters in handle (ASCII letters, digits, dashes, periods only)',
+    )
+  }
+
+  if (handle.length > 253) {
```

Comment:
The existing handle constraints are split into two sections: handle constraints & service constraints. the 40 max char total & 20 max char on each part is meant to be a handle constraint, enforced across the whole network.

Im personally still pro: "no handles over 40 chars". I just think it's bad ux & i kinda wanna enforce that everywhere we can to set a hard constraint. We already have extra constraints over domains. ie enforcing boring ascii

Do you much prefer just enforcing domain length constraints here instead?

Reply:
I don't think we are enforcing anything extra on top of domains? Punycode is still allowed in the handles themselves (which are how international domains work).

If we advertise "bring your own domain" as a feature, it does feel weird to then disallow some domains. I feel like the vast, vast majority of folks will use short handles, and we can truncate the rest at display time.

But, this is ultimately just a protocol decision and I would defer. 40 chars is fairly long.

Reply:
_/me dashes out to register a very long domain name just to troll @dholms_